openlibrary/plugins/admin/code.py

```diff
@@ -240,74 +241,73 @@ def GET(self):
-
-        if not self.is_authorized(s3_access_key, s3_secret_key):
-            raise web.unauthorized()
+    def GET(self, ocaid):
+        self.bounce_unauthorized()
 
-        # Validate input
-        i = json.loads(web.data())
-
-        if not self.validate_input(i):
-            raise web.badrequest('Missing required fields')
-
-        # Find record using OLID (raise 404 if not found)
-        edition_key = f'/books/{i.get("olid")}'
-        edition = web.ctx.site.get(edition_key)
+        i = web.input(old_ocaid=None)
+        metadata = get_metadata(ocaid)
+        if not metadata:
+            raise web.badrequest(
+                f'Failed to lookup archive.org item with ocaid {ocaid}'
+            )
+        if not metadata.get('openlibrary_edition'):
+            edition = self.get_edition_by_ocaid(i.old_ocaid or ocaid)
+            if not edition:
+                raise web.badrequest('No openlibrary_edition specified in item')
 
+        if not edition:
+            edition = web.ctx.site.get(f'/books/{metadata.get("openlibrary_edition")}')
         if not edition:
             raise web.notfound()
 
-        # Update record
-        match i.get('action', ''):
-            case 'remove':
-                self.remove_ocaid(edition)
-            case 'modify':
-                self.modify_ocaid(edition, i.get('ocaid'))
-            case '_':
-                raise web.badrequest('Unknown action')
+        if metadata.get('is_dark'):
+            self.remove_ocaid(edition)
+            return delegate.RawText(json.dumps({"status": "ok"}))
 
-        return delegate.RawText(json.dumps({"status": "ok"}))
+        # XXX : Should this occur before we fetch the metadata?  Is there a
+        # need to fetch item metadata if we're modifying the ocaid?
+        if i.old_ocaid:
+            self.modify_ocaid(edition, i.old_ocaid)
+            return delegate.RawText(json.dumps({"status": "ok"}))
 
-    def is_authorized(self, access_key, secret_key):
+        return web.badrequest('ocaid was not renamed, nor was it dark')
+
+    def bounce_authorized(self):
         """Returns True if account is authorized to make changes to records."""
-        auth = accounts.InternetArchiveAccount.s3auth(access_key, secret_key)
+        # Authenticate request:
+        s3_access_key = web.ctx.env.get('HTTP_X_S3_ACCESS', '')
+        s3_secret_key = web.ctx.env.get('HTTP_X_S3_SECRET', '')
+        auth = accounts.InternetArchiveAccount.s3auth(s3_access_key, s3_secret_key)
 
-        if not auth.get('username', ''):
-            return False
+        if auth.get('username', ''):
+            acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
+            user = acct.get_user() if acct else None
 
-        acct = accounts.OpenLibraryAccount.get(email=auth.get('username'))
-        user = acct.get_user() if acct else None
+            if user and user.is_usergroup_member('/usergroup/ia'):
```

Comment:
TODO: Populate usergroup.

Reply:
This is likely a case where we can delegate directly to a role account on Archive.org an no Open Library account is required (e.g. use xauthn to ensure that the owner of these S3 keys is e.g. `@iasync` role account)

You should be able to call the existing xauth or s3auth functions:
https://github.com/internetarchive/openlibrary/blob/master/openlibrary/accounts/model.py#L711C7-L729

Reply:
Alternatively, `admin` likely good enough for now, you decide.